postgresql_audit/base.py

```diff
@@ -446,14 +322,152 @@ class Transaction(transaction_base(base, self.schema_name)):
+                self.instrument_versioned_classes
+            ),
+            (
+                orm.mapper,
+                'after_configured',
+                self.configure_versioned_classes
+            ),
+        )
+        self.table_listeners = self.get_table_listeners()
+        self.pending_classes = WeakSet()
+        self.cached_ddls = {}
+
+    def get_table_listeners(self):
+        listeners = {'transaction': []}
+
+        listeners['activity'] = [
+            ('after_create', sa.schema.DDL(
+                render_tmpl('jsonb_change_key_name.sql', self.schema_name)
+            )),
+            ('after_create', partial(
+                    create_audit_table,
+                    schema_name=self.schema_name,
+                    use_statement_level_triggers=self.use_statement_level_triggers
+                )
+             ),
+            ('after_create', partial(create_operators, schema_name=self.schema_name))
+        ]
+        if self.schema_name is not None:
+            listeners['transaction'] = [
+                ('before_create', sa.schema.DDL(
+                    render_tmpl('create_schema.sql', self.schema_name)
+                )),
+                ('after_drop', sa.schema.DDL(
+                    render_tmpl('drop_schema.sql', self.schema_name)
+                )),
+            ]
+        return listeners
+
+    def audit_table(self, table, exclude_columns=None):
+        args = [table.name]
+        if exclude_columns:
+            for column in exclude_columns:
+                if column not in table.c:
+                    raise ImproperlyConfigured(
+                        "Could not configure versioning. Table '{}'' does "
+                        "not have a column named '{}'.".format(
+                            table.name, column
+                        )
+                    )
+            args.append(array(exclude_columns))
+        query = build_register_table_query(self.schema_name, *args)
+        if query not in cached_statements:
+            cached_statements[query] = StatementExecutor(query)
+        listener = (table, 'after_create', cached_statements[query])
+        if not sa.event.contains(*listener):
+            sa.event.listen(*listener)
+
+    def instrument_versioned_classes(self, mapper, cls):
+        """
+        Collect versioned class and add it to pending_classes list.
+
+        :mapper mapper: SQLAlchemy mapper object
+        :cls cls: SQLAlchemy declarative class
+        """
+        if hasattr(cls, '__versioned__') and cls not in self.pending_classes:
+            self.pending_classes.add(cls)
+
+    def configure_versioned_classes(self):
+        """
+        Configures all versioned classes that were collected during
+        instrumentation process.
+        """
+        for cls in self.pending_classes:
+            self.audit_table(cls.__table__, cls.__versioned__.get('exclude'))
+        assign_actor(self.base, self.transaction_cls, self.actor_cls)
```

Comment:
is this `assign_actor` call redundant or does this listener overwrite the one from the base?